ssz/src/main/java/tech/pegasys/teku/ssz/backing/collections/SszUInt64List.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.ssz.backing.collections;
+
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.ssz.backing.view.SszPrimitives.SszUInt64;
+
+public interface SszUInt64List extends SszPrimitiveList<UInt64, SszUInt64> {
```

Comment:
question - are we actually getting any real benefit from having these interfaces separate to the implementations?  If there are multiple implementations or the interfaces are narrowing the exposed API then keeping them is fine, but otherwise things would be a lot simpler if we just inlined the interfaces and used the implementations directly.

Reply:
It looks quite difficult to pick the interfaces apart and I think there may be some multiple inheritance going on which may mean they are giving value.

Reply:
> question - are we actually getting any real benefit from having these interfaces separate to the implementations?

Actually no benefits just except of a shorter notation. I.e. `SszUInt64List` instead of `SszPrimitiveList<UInt64, SszUInt64>`. That's what we were chatting about here https://github.com/Nashatyrev/artemis/pull/11 

I had stick to your proposal below: 
> I would be very tempted to introduce classes for quite commonly used types (SszBytes32Vector for example). Not sure I'd do it for all of them and would probably just use generics for the mutable version.

TBH I'm also not a big fan of that variety of specific collection as well as don't like too much extra-verbose `SszPrimitiveList<UInt64, SszUInt64>`. So this was a kind of trade-off 

The root reason of why we couldn't just write `SszPrimitiveList<UInt64>`or even `SszList<UInt64>` is they are all extends 
`SszComposite<SszChildT extends SszData>` and collection elements should extend `SszData`. 
I was trying the case `SszComposite<SszChildT>` (i.e. the generic element type) but that way we would loose a lot of type safety.

So I'm honestly not sure what's the best solution here

Reply:
I definitely prefer having `SszUInt64List` as effectively an alias for `SszPrimitiveList<UIn64, SszUInt64>`.
I was more meaning the whole interface hierarchy, not just this one particular one. So essentially use `SszUInt64ListImpl` and `SszPrimitiveListImpl` everywhere the interface currently is.  There are a bunch of classes with quite a direct naming like that (possibly not those two - just an example) and often when you see classes with names ending in `Impl` there's not actually any value to having the interface separate.

I think having dug through the details some more, I'm satisfied that the interfaces are worth having.

Reply:
My primary intention was to expose pure interface 'API'. From my perspective complex hierarchies are far better understandable when based on just interfaces than on mix of interfaces/impl classes

Basically my personal preference (last years) is to create an interface and then do implementation class for more or less complex classes (even when just a single implementation is intended originally). It makes both developing and later reading the code more clear from my perspective